standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/RetryHandler.java

```diff
@@ -0,0 +1,299 @@
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.dao.DataAccessException;
+import org.springframework.jdbc.UncategorizedSQLException;
+import org.springframework.transaction.TransactionStatus;
+import org.springframework.transaction.support.TransactionTemplate;
+
+import java.sql.SQLException;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Pattern;
+
+public class RetryHandler {
+
+  @FunctionalInterface
+  public interface VoidRetryFunction {
+
+    void call(TransactionStatus status) throws SQLException, MetaException;
+
+  }
+
+  @FunctionalInterface
+  public interface RetryFunction<Result> {
+
+    Result call(TransactionStatus status) throws SQLException, MetaException;
+
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(RetryHandler.class);
+
+  private static final int ALLOWED_REPEATED_DEADLOCKS = 10;
+  private static final String MANUAL_RETRY = "ManualRetry";
+
+  /**
+   * Derby specific concurrency control
+   */
+  private static final ReentrantLock derbyLock = new ReentrantLock(true);
+
+
+  private DatabaseProduct dbProduct;
+
+  private long deadlockRetryInterval;
+  private long retryInterval;
+
+  private int deadlockCnt;
+  private int retryLimit;
+  private int retryNum;
+  
+  private Configuration conf;
+
+
+  public static String getMessage(Exception ex) {
+    StringBuilder sb = new StringBuilder();
+    sb.append(ex.getMessage());
+    if (ex instanceof SQLException) {
+      SQLException sqlEx = (SQLException)ex; 
+      sb.append(" (SQLState=").append(sqlEx.getSQLState()).append(", ErrorCode=").append(sqlEx.getErrorCode()).append(")");
+    }
+    return sb.toString();
+  }  
+  
+  public void init(Configuration conf, DatabaseProduct dbProduct){
+    this.conf = conf;
+    this.dbProduct = dbProduct;
+
+    retryInterval = MetastoreConf.getTimeVar(conf, MetastoreConf.ConfVars.HMS_HANDLER_INTERVAL,
+        TimeUnit.MILLISECONDS);
+    retryLimit = MetastoreConf.getIntVar(conf, MetastoreConf.ConfVars.HMS_HANDLER_ATTEMPTS);
+    deadlockRetryInterval = retryInterval / 10;
+  }
+  
+  public <Result> Result executeWithRetry(CallProperties properties, RetryFunction<Result> function) throws MetaException {
+    return executeWithRetryInternal(properties, function, true);
+  }
+
+  public void executeWithRetry(CallProperties properties, VoidRetryFunction function) throws MetaException {
+    executeWithRetryInternal(properties, (TransactionStatus status) -> { 
+      function.call(status); 
+      return null;
+    }, true);
+  }
+
+  public  <Result> Result executeWithoutRetry(CallProperties properties, RetryFunction<Result> function) throws MetaException {
+    return executeWithRetryInternal(properties, (TransactionStatus status) -> {
+      function.call(status);
+      return null;
```

Comment:
Should the return value of the function be returned instead if `null`?

Reply:
Good catch, this was a copy-paste issue.